Service Fee (Pricing) documentation (#2179)
nucypher/nucypher: docs/source/architecture/servicefees.rst

.. _service-fees:

Service Fees (Pricing)
======================

Minimum fee rate
----------------

> When they join the network, stakers choose a minimum fee rate, on a *per sharing policy* and *per 24h period* basis, that their worker machine (Ursula) will accept at the point of engagement with a network user (Alice). If Alice’s offered rate and commensurate deposit (contained alongside the sharing policy’s parameters in an ``Arrangement`` object), for a specified policy duration, computes as equal to or greater than the minimum fee rate, the sharing policy will be accepted and the access control service will commence. Ursula will service the sharing policy by being online and answering access requests, at that unchanging fee rate, until the specified expiration date or an early revocation instigated by Alice.

Author: Agreed, mentioning the deposit here is unnecessarily confusing!

> The minimum fee rate is also calculated *per Ursula*. If Alice wishes to employ multiple Ursulas to service a single sharing policy (``n`` > 1), a common configuration, then they must pay each staker associated with that policy the same fee rate. Although Alices may attempt price optimization strategies to find the cheapest group of Ursulas, they will nevertheless have to pay the highest rate from the set of minimum fee rates from the stakers with which they end up engaging.

Author: I agree that this needs rewording! I moved the optimization to the end. 
This paragraph is basically introducing 2 concepts: that the rate is per-machine, and that Alices pay the highest of the set of specified min rates. I think the 'Even though' mixes these concepts together 

```suggestion
The minimum fee rate applies to each individual worker machine managing a given sharing policy. In other words, the rate is also *per Ursula*. 

If Alice wishes to employ multiple Ursulas to service a single sharing policy (``n`` > 1), a common configuration, then they must pay each staker associated with that policy the same exact fee rate. This rate will be the highest from the set of *minimum* fee rates specified by the stakers with which they engage. Alices may attempt price optimization strategies to find the cheapest group of Ursulas.
```

> The minimum fee rate is also calculated *per Ursula*. If Alice wishes to employ multiple Ursulas to service a single sharing policy (``n`` > 1), a common configuration, then they must pay each staker associated with that policy the same fee rate. Although Alices may attempt price optimization strategies to find the cheapest group of Ursulas, they will nevertheless have to pay the highest rate from the set of minimum fee rates from the stakers with which they end up engaging.

Reviewer: 👍

> The minimum fee rate is also calculated *per Ursula*. If Alice wishes to employ multiple Ursulas to service a single sharing policy (``n`` > 1), a common configuration, then they must pay each staker associated with that policy the same fee rate. Although Alices may attempt price optimization strategies to find the cheapest group of Ursulas, they will nevertheless have to pay the highest rate from the set of minimum fee rates from the stakers with which they end up engaging.

Reviewer: One comment in your updated edit "same exact" -> "same"

Alices are required to escrow a deposit covering the cost of the entire duration of the policy, but fees are paid out to stakers once per period, after their Ursula confirms activity. Therefore the minimum sum of fees a staker will receive each period is the product of their specified minimum fee rate and the number of active sharing policies their Ursula has been assigned.

Global fee range
----------------

> The global fee range is a means of establishing quasi-universal pricing for the NuCypher service. It is enforced via the function ``feeRateRange`` in ``PolicyManager.sol``, which specifies constraints, expressed in GWEI, per sharing policy and per 24h period. The minimum fee rate rate chosen by stakers must fall within the global fee range. The network will launch with the following parameters:

Author: @vzotova Nice spot! I will change it to WEI here. Even so, I think elsewhere, it's more intelligible to quote/discuss prices in GWEI – e.g. Default fee rate = 200 GWEI (~$4.47 x10^-5) per period. Wdyt?

**Minimum fee rate**

XX GWEI (~$X.XX x10^-X) *per period*

XX,XXX GWEI ($X.XXX) *per year*

**Maximum fee rate**

X,XXX GWEI (~$X.XX x10^-X) *per period*

XXX,XXX GWEI ($X.XXX) *per year*

**Default fee rate**

XXX GWEI (~$X.XX x10^-X) *per period*

XX,XXX GWEI ($X.XXXX) *per year*

*1 GWEI = 10^-9 ETH*

> *USD conversion utilizes the ETHUSD 100-day rolling average of 1 ETH = $222 (08.08.20)*

Author: The exchange rate is used here to estimate the per period and per year earnings for each sharing policy in fiat – the figures in parentheses. It's quite useful to know the max rate per policy per year is $0.081, for example.

Also the pricing paper we use customer budgets, staker operational costs and other estimates that are denominated in USD, as an input to analysis and calculations. I agree that it's clunky and impermanent. Would be cool if we could update these $ rates via the CMC API so it's always accurate

> Note that the minimum and maximum fee rate are an upper and lower bound to constrain the fee rate a staker may offer. The default fee rate is the rate that will be displayed and offered to Alices if the staker chooses not to configure this parameter themselves.

Author: @vzotova Two perfectly plausible scenarios, I'll add them!

The fee range must be adhered to in identical fashion by all NuCypher stakers, regardless of their stake size or capacity. The fee range applies to all sharing policies, irrespective of the volume of re-encryption requests or other distinguishing attributes. It also applies equally to all periods in the future, until the moment that the global fee range’s parameters are adjusted or the range is removed, via official governance processes (see below). If an update of this sort occurs, sharing policies that were previously established, but have not yet expired, should not have the per-period fee rate retroactively modified.

Governance & pricing paper
---------------------------------------

> In order to successfully execute the ``PolicyManager.sol`` contract, the global fee range must be adhered to by the Ursula (and Alice). Failing this, the contract will throw up an error and it will not be possible to commence a commercial engagement or pay/receive fees. Attempts to circumvent NuCypher’s smart contracts are likely to be futile (except perhaps with limited, existing, trust-heavy relationships), given the requirement of coordinated modification and redeployment by network users and a critical mass of other stakers.

Reviewer: execute -> interact 
removed `(except perhaps with limited, existing, trust-heavy relationships)` -> should we mention this?
```suggestion
In order to successfully interact with the ``PolicyManager.sol`` contract, the global fee range must be adhered to by the Ursula (and Alice). Failing this, the contract will throw up an error and it will not be possible to commence a commercial engagement or pay/receive fees. Attempts to circumvent NuCypher’s smart contracts are likely to be futile given the requirement of coordinated modification and redeployment by network users and a critical mass of other stakers.
```

Author: I was on the fence about this – it's good to acknowledge extra-protocol interactions, rather than pretend they don't exist. At the same time perhaps saying "circumventing is futile" is taking too strong a stance on it. The main point is that you can't commercially engage unless you adhere to the fee range

> Given its high enforceability, the presence of an inflexible fee range dictating the bounds of every transaction is arguably the most critical component of the NuCypher protocol’s economic design and parametrization, particularly over the long-term and with respect to the sustainability of the network. From a governance perspective, it is also amongst the most malleable. If a quorum of stakers wish to set prices outside the range, then they have the right to lobby and propose a widening of the global fee range, its removal altogether, or some other design modification (e.g. narrowing the range). They may do so via the NuCypher DAO – the owner of all NuCypher smart contracts, and controlled by stakers in proportion to their stake size. See <LINK> for guidance on the NuCypher DAO and official NuCypher governance processes.

Reviewer: all NuCypher smart contracts, and controlled by stakers in proportion to their stake size. -> all NuCypher smart contracts -  by submitting a proposal which is voted on by stakers weighted in proportion to their stake size.
 
```suggestion
Given its high enforceability, the presence of an inflexible fee range dictating the bounds of every transaction is arguably the most critical component of the NuCypher protocol’s economic design and parametrization, particularly over the long-term and with respect to the sustainability of the network. From a governance perspective, it is also amongst the most malleable. If a quorum of stakers wish to set prices outside the range, then they have the right to lobby and propose a widening of the global fee range, its removal altogether, or some other design modification (e.g. narrowing the range). They may do so via the NuCypher DAO – the owner of all NuCypher smart contracts -  by submitting a proposal which is voted on by stakers weighted in proportion to their stake size. See <LINK> for guidance on the NuCypher DAO and official NuCypher governance processes.
```

The **Pricing Protocol & Economics paper** serves as a key resource and reference for community debate, proposals for modification and DAO-driven upgrades and redeployments in the future. The paper discusses the merits and risks of quasi-universal pricing and the enforcement of an upper and lower bound on all offered price points. It includes a price point analysis from a demand-side, service-side and theoretical standpoint to produce the constraints in absolute terms (i.e. in GWEI) that the network will launch with.

Setting a discretionary fee rate
--------------------------------

Stakers should use the ``setMinFeeRate`` function to specify the minimum fee rate that their Ursula (worker machine) will accept.

> Note that Alices seeking to instantiate a new sharing policy are able to first discover all current minimum fee rates available to them, by retrieving the list of active stakers’ addresses, then querying the public variable ``PolicyManager.nodes(staker_address).minRewardRate`` with each ``staker_address``.

Reviewer: ```suggestion
Note that Alices seeking to instantiate new sharing policies are able to first discover all current minimum fee rates available to them, by retrieving the list of active stakers’ addresses, then querying the public variable ``PolicyManager.nodes(staker_address).minRewardRate`` with each ``staker_address``.
```

Author: Ah, so `PolicyManager.nodes(staker_address).minRewardRate` would return values from stakers who've (naughtily) specified rates outside the range?

Reviewer 2: To set `PolicyManager.nodes(staker_address).minRewardRate`  value - staker must choose within global range, but if global range changes - then `minRewardRate` could be outdated, also before staker even set this value - it'll be zero

Setting a price point, even within a tight range, requires the evaluation and weighting of many factors against one another. Many of these considerations are unique to the staker, such as their ongoing operational costs, economy of scale (e.g. through participation in other networks) and participation timeframe. However, the most important factors to consider pertain to the holistic service from the perspective of network users – for example, the affordability, congruency, and stability, of all offered price points – i.e. how probable it is that prices remain affordable to a developer after they are irreversibly committed to integrating NuCypher access control into their application’s technology stack.

For an overview of price setting considerations, see the *Pricing Strategies* section of the Pricing Protocol & Economics paper.

Operational costs
-----------------

> The cost of operating a typical Ursula, at network genesis, is estimated to be between $X and $Y per month. The variability of these estimates is driven primarily by diverse infrastructure costs across geographical locations, the range of feasible strategies for minimizing gas costs, and the economies of scale associated with service provision in multiple decentralized networks. This does not include the risks and opportunity costs of locking the Nu token for an extended duration of time. For a full derivation of overhead scenarios and the underlying assumptions, see the *Service-driven Pricing* section of the Pricing Protocol & Economics paper.

Reviewer: Nu -> NU
```suggestion
The cost of operating a typical Ursula, at network genesis, is estimated to be between $X and $Y per month. The variability of these estimates is driven primarily by diverse infrastructure costs across geographical locations, the range of feasible strategies for minimizing gas costs, and the economies of scale associated with service provision in multiple decentralized networks. This does not include the risks and opportunity costs of locking the NU token for an extended duration of time. For a full derivation of overhead scenarios and the underlying assumptions, see the *Service-driven Pricing* section of the Pricing Protocol & Economics paper.
```

The minimum fee rate applies to each individual worker machine managing a given sharing policy. In other words, the rate is also *per Ursula*. 

If Alice wishes to employ multiple Ursulas to service a single sharing policy (``n`` > 1), a common configuration, then they must pay each staker associated with that policy the same exact fee rate. This rate will be the highest from the set of *minimum* fee rates specified by the stakers with which they engage. Alices may attempt price optimization strategies to find the cheapest group of Ursulas.

> When issuing a sharing policy, Alices are required to escrow a deposit covering the cost of the entire duration of the policy. The deposit is paid out in tranches to stakers once per period, in the period after their Ursula makes a commitment (formerly referred to as *confirm activity*). Therefore, the minimum sum of fees a staker can receive each period is the product of their specified minimum fee rate and the number of active sharing policies their Ursula has been assigned.

Reviewer 2: *in the period after their Ursula makes a commitment* -> slightly wrong
1st period - Ursula makes a commitment
2nd period -> this period
3rd period -> staker can get fee for policy for 2nd period

Author: ```suggestion
When issuing a sharing policy, Alices are required to escrow a deposit covering the cost of the entire duration of the policy. This deposit is split and paid out in tranches to stakers once per period. 

The precise payment flow follows a repeated three-period cycle; in the first period the Ursula *makes a commitment* to the second period. Then, the Ursula services the policy during the second period (and makes a commitment to the third period). In the third period, they receive the fee tranche for their work during the second period, and continue servicing the policy, etc. This cycle continues until all the policies that Ursula is responsible for have expired.

The minimum sum of fees a staker can receive period-to-period is the product of their specified minimum fee rate and the number of active sharing policies their Ursula has been assigned. 

Note that *making a commitment* was formerly referred to as *confirming activity*.
```

Global fee range
----------------

The global fee range is a means of establishing quasi-universal pricing for the NuCypher service. It is enforced via the function ``feeRateRange`` in ``PolicyManager.sol``, which specifies per sharing policy and per 24h period constraints expressed in **WEI**. In other material, rates are discussed in **GWEI** (and fiat).

The minimum fee rate chosen by stakers must fall within the global fee range. The network will launch with the following parameters:

**Minimum fee rate**

XX GWEI (~$X.XX x10^-X) *per period*

XX,XXX GWEI ($X.XXX) *per year*

> **Maximum fee rate**

Reviewer 3: The term "maximum fee rate" seems to imply it's an absolute upper bound to the fee rate, while what we have is an upper bound on the _offered_ rate. Or in other words, nothing prevents Alice from paying a price higher than this rate; this may be particularly relevant with handpicked Ursulas that offer a premium service.

Author: @vzotova warned me about this a few months ago and it slipped my mind 😏 
We can add a caveat and change the naming slightly. I do believe this is a quite extreme edge case tho 

- As it stands, any rate specified by a staker that's >max will be changed and shown to Alices as the default rate (in my draft pricing, this figure is quite a lot lower). 
- They then have to accept that rate for their service for any customer that comes along, regardless of any premium feature or differentiating attribute, and regardless of the existence of other Alices that are willing to pay more 
- So within the current rules, if you think there are Alices willing to pay 'more than normal' for your specific Ursula, it makes more sense to choose the max permitted rate and/or lobby for a widening of the range.

On Alice's side, while it's possible to posit a >max rate in their Arrangement, there are some reasons why they are unlikely to do so within 'normal' behaviour.
- One cannot easily identify Ursulas with premium services using their specified rate alone, unless they equal or are close to the max
- If stakers are restrained by the fee range, paying them more than needed is probably part of a rare strategy or trustful/dodgy relationship (e.g. cycling money around to keep that particular staker running) 
- In general, requiring multiple Ursulas for a single policy could make benefitting from unique services a high friction exercise

Author: ```suggestion
**Maximum chargeable fee rate**
```

Note that the minimum and maximum fee rate are a lower and upper bound to constrain the fee rate a staker may offer. The default fee rate is the rate that will be displayed and offered to Alices if the staker chooses not to configure this parameter themselves, or chooses a rate outside the boundaries of the global fee range. The default rate will also be used if the range's boundaries are updated, a staker's specified rate *now* falls outside the range, and they fail to change it.

The fee range must be adhered to in identical fashion by all NuCypher stakers, regardless of their stake size or capacity. The fee range applies to all sharing policies, irrespective of the volume of re-encryption requests or other distinguishing attributes. It also applies equally to all periods in the future, until the moment that the global fee range’s parameters are adjusted or the range is removed, via official governance processes (see below). If an update of this sort occurs, sharing policies that were previously established, but have not yet expired, should not have the per-period fee rate retroactively modified.

> 

Author: ```suggestion
Note that the global fee range is only applicable to stakers and Ursulas. Alices are free to pay as high a rate as they like.
```
@cygnusv

.. _service-fees:

Service Fees (Pricing)
======================

Minimum fee rate
----------------

> When they join the network, stakers specify a minimum fee rate, on a *per sharing policy* and *per 24h period* basis, that their worker (Ursula) will accept at the point of engagement with a network user (Alice). If Alice’s offered per period rate (contained alongside the sharing policy’s parameters in an ``Arrangement`` object) for a specified policy duration computes as equal to or greater than the minimum fee rate, the sharing policy will be accepted by Ursula and the access control service will commence. Ursula will service the sharing policy by being online and answering access requests, at that unchanging per period fee rate, until the specified expiration date or an early revocation is instigated by Alice.

Reviewer 4: ```suggestion
When stakers join the network, they specify a minimum fee rate, on a *per sharing policy* and *per 24h period* basis, that their worker (Ursula) will accept at the point of engagement with a network user (Alice). If Alice’s offered per period rate (contained alongside the sharing policy’s parameters in an ``Arrangement`` object) for a specified policy duration computes as equal to or greater than the minimum fee rate, the sharing policy will be accepted by Ursula and the access control service will commence. Ursula will service the sharing policy by being online and answering access requests, at that unchanging per period fee rate, until the specified expiration date or an early revocation is instigated by Alice.
```

The minimum fee rate applies to each individual worker machine managing a given sharing policy. In other words, the rate is also *per Ursula*. 

> If Alice wishes to employ multiple Ursulas to service a single sharing policy (``n`` > 1), a common configuration, then they must pay each staker associated with that policy the same exact fee rate. This rate will be the highest from the set of *minimum* fee rates specified by the stakers with which they engage. Alices may attempt price optimization strategies to find the cheapest group of Ursulas.

Reviewer: ```suggestion
If Alice wishes to employ multiple Ursulas to service a single sharing policy (``n`` > 1), a common configuration, then they must pay each staker associated with that policy the same fee rate. This rate will be the highest from the set of *minimum* fee rates specified by the stakers with which they engage. Alices may attempt price optimization strategies to find the cheapest group of Ursulas.
```

The precise payment flow follows a repeated three-period cycle; in the first period the Ursula *makes a commitment* to the second period. Then, the Ursula services the policy during the second period (and makes a commitment to the third period). In the third period, they receive the fee tranche for their work during the second period, and continue servicing the policy, etc. This cycle continues until all the policies that Ursula is responsible for have expired.

The minimum sum of fees a staker can receive period-to-period is the product of their specified minimum fee rate and the number of active sharing policies their Ursula has been assigned. 

Note that *making a commitment* was formerly referred to as *confirming activity*.

Global fee range
----------------

The global fee range is a means of establishing quasi-universal pricing for the NuCypher service. It is enforced via the function ``feeRateRange`` in ``PolicyManager.sol``, which specifies per sharing policy and per 24h period constraints expressed in **WEI**. In other material, rates are discussed in **GWEI** (and fiat).

> The minimum fee rate chosen by stakers must fall within the global fee range. The network will launch with the following parameters:

Reviewer 4: Should we mention this can be changed by the DAO?
```suggestion
The minimum fee rate chosen by stakers must fall within the global fee range. This global fee range is set collectively by stakers via the NuCypher DAO. The network will launch with the following parameters:
```

Reviewer 4: I see that you address this later so you can disregard this comment.

> **Minimum fee rate**

Author: ```suggestion
**Minimum chargeable fee rate**
```

The fee range must be adhered to in identical fashion by all NuCypher stakers, regardless of their stake size or capacity. The fee range applies to all sharing policies, irrespective of the volume of re-encryption requests or other distinguishing attributes. It also applies equally to all periods in the future, until the moment that the global fee range’s parameters are adjusted or the range is removed, via official governance processes (see below). If an update of this sort occurs, sharing policies that were previously established, but have not yet expired, should not have the per-period fee rate retroactively modified.
Note that the global fee range is only applicable to stakers and Ursulas. Alices are free to pay as high a rate as they like.

> Governance & pricing paper

Reviewer 4: ```suggestion

Governance & pricing paper
```

Note that Alices seeking to instantiate new sharing policies are able to first discover all current minimum fee rates available to them, by retrieving the list of active stakers’ addresses, then querying the public variable ``PolicyManager.getMinFeeRate(staker_address)`` with each ``staker_address``.

Setting a price point, even within a tight range, requires the evaluation and weighting of many factors against one another. Many of these considerations are unique to the staker, such as their ongoing operational costs, economy of scale (e.g. through participation in other networks) and participation timeframe. However, the most important factors to consider pertain to the holistic service from the perspective of network users – for example, the affordability, congruency, and stability, of all offered price points – i.e. how probable it is that prices remain affordable to a developer after they are irreversibly committed to integrating NuCypher access control into their application’s technology stack.

For an overview of price setting considerations, see the *Pricing Strategies* section of the Pricing Protocol & Economics paper.

Operational costs
-----------------

> The cost of operating a typical Ursula, at network genesis, is estimated to be between $X and $Y per month. The variability of these estimates is driven primarily by diverse infrastructure costs across geographical locations, the range of feasible strategies for minimizing gas costs, and the economies of scale associated with service provision in multiple decentralized networks. This does not include the risks and opportunity costs of locking the NU token for an extended duration of time. For a full derivation of overhead scenarios and the underlying assumptions, see the *Service-driven Pricing* section of the Pricing Protocol & Economics paper.

Reviewer 4: "the NU token" -> "NU"

```suggestion
The cost of operating a typical Ursula, at network genesis, is estimated to be between $X and $Y per month. The variability of these estimates is driven primarily by diverse infrastructure costs across geographical locations, the range of feasible strategies for minimizing gas costs, and the economies of scale associated with service provision in multiple decentralized networks. This does not include the risks and opportunity costs of locking NU for an extended duration of time. For a full derivation of overhead scenarios and the underlying assumptions, see the *Service-driven Pricing* section of the Pricing Protocol & Economics paper.
```

Note on staker sustainability
-----------------------------

> Although the maximum fee rate parameter constrains the income from fees in one plane, it is a component of a strategy to maximize long-term network revenue through predictable, affordable and congruent pricing. Operational costs will almost certainly exceed fee income in the near-term, but the subsidy mechanism is designed to steadily support service-providers for the first 5 to 8 years – see the *Demand uncertainty & fragility* section of the Staking & Economic Protocol paper for more detail. This stable source of income provides an extended window for the NuCypher community to trial various fee range parameters until a balance is struck between the extremes of 1) unaffordability for early customers leading to low demand, and 2) unsustainability for service-providers leading to low participation.

Reviewer 4: ```suggestion
Although the maximum fee rate parameter constrains the income from fees in one plane, it is a component of a strategy to maximize long-term network revenue through predictable, affordable and congruent pricing. Operational costs will almost certainly exceed fee income in the near-term, but the subsidy mechanism is designed to steadily support service-providers for the first 5 to 8 years – see the *Demand uncertainty & fragility* section of the Staking & Economic Protocol paper for more detail. This subsidy provides an extended window for the NuCypher community to trial various fee range parameters until a balance is struck between the extremes of 1) unaffordability for early customers leading to low demand, and 2) unsustainability for service-providers leading to low participation.
```
